src/lib/y2partitioner/widgets/btrfs_devices.rb

```diff
@@ -0,0 +1,228 @@
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "yast2/popup"
+require "cwm"
+require "y2partitioner/widgets/btrfs_metadata_raid_level"
+require "y2partitioner/widgets/btrfs_data_raid_level"
+require "y2partitioner/widgets/btrfs_devices_selector"
+
+module Y2Partitioner
+  module Widgets
+    # Widget grouping a set of widgets related to the selection of Btrfs devices
+    #
+    # It contains widgets to select the metadata/data RAID levels and also to select the devices
+    # used by the filesystem.
+    class BtrfsDevices < CWM::CustomWidget
+      # Constructor
+      #
+      # @param controller [Actions::Controllers::BtrfsDevices]
+      def initialize(controller)
+        textdomain "storage"
+
+        @controller = controller
+      end
+
+      # @macro seeCustomWidget
+      def contents
+        VBox(
+          Left(
+            HVSquash(
+              HBox(
+                metadata_raid_level_widget,
+                HSpacing(1),
+                data_raid_level_widget
+              )
+            )
+          ),
+          VSpacing(1),
+          btrfs_devices_selector_widget
+        )
+      end
+
+      # @macro seeCustomWidget
+      def help
+        help_for_raid_levels + help_for_default_raid_level
+      end
+
+      # @macro seeAbstractWidget
+      #
+      # Validates the selected values
+      #
+      # An error popup is shown when there are errors in the selected values.
+      #
+      # @return [Boolean]
+      def validate
+        current_errors = errors
+        return true if current_errors.none?
+
+        message = current_errors.join("\n\n")
+        Yast2::Popup.show(message, headline: :error)
+
+        false
+      end
+
+    private
+
+      # @return [Actions::Controllers::BtrfsDevices]
+      attr_reader :controller
+
+      # Widget to select the metadata RAID level
+      #
+      # @return [Widgets::BtrfsMetadataRaidLevel]
+      def metadata_raid_level_widget
+        @metadata_raid_level_widget ||= Widgets::BtrfsMetadataRaidLevel.new(controller)
+      end
+
+      # Widget to select the data RAID level
+      #
+      # @return [Widgets::BtrfsDataRaidLevel]
+      def data_raid_level_widget
+        @data_raid_level_widget ||= Widgets::BtrfsDataRaidLevel.new(controller)
+      end
+
+      # Widget to select devices used by the filesystem
+      #
+      # @return [Widgets::BtrfsDevicesSelector]
+      def btrfs_devices_selector_widget
+        @btrfs_devices_widget ||= Widgets::BtrfsDevicesSelector.new(controller)
+      end
+
+      # Btrfs RAID levels help
+      #
+      # @return [String]
+      def help_for_raid_levels
+        _("<h4><b>Btrfs RAID levels</b></h4>" \
+          "<p> Btrfs supports the following RAID levels for both, metadata and data:" \
```

Comment:
Any reason for the white space right after `<p>`?

Reply:
Leftover.